src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -414,5 +432,95 @@ private static void CheckForCompletedTransfers(SafeCurlMultiHandle multiHandle,
                 }
             }
         }
+
+        private async static void PollFunction(SafeCurlMultiHandle multiHandle)
+        {
+            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+            while (true)
+            {
+                if (cancellationTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                int count = 0;
+                lock (multiHandle)
+                {
+                    multiHandle.CancellationTokenSource = cancellationTokenSource;
+                    if (null != multiHandle.FdSet)
+                    {
+                        count = multiHandle.FdSet.Count;
+                    }
+                }
+                if (count == 0)
+                {
+                    // This can happen rarely when the socket callback is not
+                    // yet invoked for the first request
+                    await Task.Delay(100).ConfigureAwait(false);
+                    continue;
+                }
+
+                int result = -1;
+                unsafe
+                {
+                    Interop.libc.pollfd* fds = stackalloc Interop.libc.pollfd[count];
+                    lock (multiHandle)
+                    {
+                        if (count != multiHandle.FdSet.Count)
+                        {
+                            continue;
+                        }
+                        int i = 0;
+                        foreach (int fd in multiHandle.FdSet)
+                        {
+                            fds[i].fd = fd;
+                            fds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                            i++;
+                        }
+                    }
+
+                    var retVal = Interop.libc.poll(fds, (uint)count, -1);
```

Comment:
I think I'm still unclear as to the overall design here, why/when polling with the timer/async is needed, why/when we instead choose to block for an unbounded period of time (as in this case).  Can you elaborate on the design here, including why whatever mechanism libcurl provides is insufficient and that leads us to need to do this?

From what I see, we're maintaining a collection of FDs, one per operation being performed concurrently on the HttpClient/HttpClientHandler.  We want to be notified when some sort of event occurs on one of those FDs.  That's the fundamental issue trying to be addressed here, right?  And libcurl doesn't provide appropriate notifications for that?  

It looks like we're setting up a polling loop that continually fires every 100ms until at least one FD is available, and then we block indefnitely waiting for something to happen on one of those available FDs.  That seems problematic for a few reasons:
1. We're doing work every 100ms even if there's nothing to be done.  I'd expect if there are no FDs registered, there'd be no need to have the loop running.
2. 100ms is seems to me like a fairly long time to not process data that might be available.
2. We're then blocking waiting for work on one of the registered FDs.  What if concurrently with that waiting another operation kicks off... its data could be available, but we won't process it because we're not waiting on it in this polling operation, right?

I'm trying to understand the lay of the land and this particular design before we come up with an alternative.